### DISCUSSION TITLE
[scenario] Anticipating port conflicts with a local active server queue on a shared computer

### DISCUSSION THREAD

**Discussion by wlandau** [r-lib/mirai#26] [Q&A] [answered]
My colleagues and I work with a traditional SGE cluster which has a small handful of login nodes and several compute nodes. Everyone shares the nodes, particularly the login nodes, but even the compute nodes (even when specific cores are reserved).

Let's say Alice and Bob each log into the login node with IP address 192.168.0.2 on the cluster's local network. Each uses a [local active server queue](https://github.com/shikokuchuo/mirai#connecting-to-remote-servers-through-a-local-server-queue) to run a large pipeline.

Alice listens for servers on client ports 50000 through 50999 and launches 1000 servers to dial into them:

```r
daemons("tcp://192.168.0.2:50000", nodes = 1000) 
```

Similarly, Bob listens for servers on client ports 50050 through 51049 and launches 1000 servers to dial into them:

```r
daemons("tcp://192.168.0.2:50050", nodes = 1000) 
```

Alice and Bob both launch tasks which go to servers that dial into port 50100.

```r
mirai_alice <- mirai("Alice's task")
```

```{r}
mirai_bob <- mirai("Bob's task")
```

What will `mirai_alice$data` return on Alice's client? Will it return `"Alice's task"` or `"Bob's task"`? How do the servers know which client should get the result? (This is what I meant by "crossed wires" in https://github.com/shikokuchuo/mirai/discussions/24.)

**Comment by wlandau** [comment]
I anticipate similar situations when a `mirai` user likes to use port 8888 but it conflicts with someone else's Jupyter notebook running from the same port on the same login node of the cluster. That is the main reason why I would prefer to check `parallelly::freePort()` and then only use that port. But maybe I am missing something about how TCP sockets work.

**Reply by shikokuchuo** [reply]
I had a brief look at `parallelly::freePort()` and it seems to allocate a random port, so fundamentally should not be any better than asking NNG to assign a free port (by specifying the port as 0).

As the client is using listeners, they actually bind to the requested port. So if it is already in use you are sure to get an error 10 - address in use or 16 - permission denied (say you if you try listening to SSH port 22).

**Comment by shikokuchuo** [comment, marked as answer]
There is no way for both Bob and Alice to listen on the same port.

If they run nodes directly they will get:

Bob:
```r
daemons("tcp://192.168.0.2:50000") 
#> [1] "remote"
```
Alice:
```r
daemons("tcp://192.168.0.2:50000") 
#> Error in socket(protocol = "req", listen = value) : 10 | Address in use
```
If they add the nodes argument, they don't see the above error as it is the background process that listens.

However for Alice, her background queue process errors and exits. The easiest way of detecting this is simply querying `daemons()`, which will show 0 connections instead of 1.

**Reply by wlandau** [reply]
Thanks for explaining. It looks like I will have to be about as careful with client ports as I anticipated. But I could see many cloud-based setups that will not run into this issue.

**Reply by shikokuchuo** [reply]
Are you concerned about the contiguous range? That was just an abstraction I added so we only have to supply one port number. If you want the option to specify arbitrary port numbers (that you have checked are available) let me know I can add that as an option.

**Reply by wlandau** [reply]
Thanks, I think relaxing the contiguous range would definitely help if I go that direction. But it still sounds like it would be complicated for both of us. Because of that, the difficulty of autoscaling up and down, and the potential for port conflicts, I would first like to spend time revisiting paths forward based on `mirai`'s passive queue.

**Reply by wlandau** [reply]
I looked back at https://github.com/shikokuchuo/mirai/discussions/22#discussioncomment-5121947, and after more testing, I think I understand more about the limitations of the passive queue. In my test (see below), I tried to implement a somewhat active queue of my own: a client loop that avoids calling `mirai()` until one of the servers is definitely available. Unfortunately, it looks like NNG's round-robin algorithm still chooses to submit to the busy server and ignore the free one.

With the active queue, it seems like the way forward involves a lot of work for both of us. Up-scaling servers after a down-scale as in #25, and dealing with port conflicts on shared machines in #26, still seem like tough obstacles even after your helpful comments.

```{r}
# Client
library(mirai)
library(purrr)
daemons("tcp://127.0.0.1:55001")
run_task <- function(index) {
  seconds <- 19 * (index %% 2) + 1
  Sys.sleep(seconds)
  paste0("task=", index, " pid=", ps::ps_pid(), " seconds=", seconds)
}

# Two servers
library(mirai)
server("tcp://127.0.0.1:55001")

# Client loop
start <- proc.time()["elapsed"]
queue <- seq_len(3)
mirais <- list()
while(TRUE) {
  # Submit available tasks to all available servers.
  while (length(queue) > 0 && length(mirais) < daemons()$connections) {
    mirais <- c(mirais, mirai(run_task(index), run_task = run_task, index = queue[1]))
    queue <- queue[-1]
  }
  # Grab any available results.
  done <- integer(0)
  for (index in seq_along(mirais)) {
    if (!unresolved(mirais[[index]])) {
      done <- c(done, index)
      finished <- round(proc.time()["elapsed"] - start, 2)
      cat(mirais[[index]]$data, " finished=", finished, "\n", sep = "")
    }
  }
  if (length(done)) mirais <- mirais[-done]
  # Exit if all done.
  if (length(queue) == 0 && length(mirais) == 0) {
    break
  }
  Sys.sleep(1e-3)
}

# Output on the client:
# task=2 pid=31874 seconds=1 finished=1.04
# task=1 pid=31864 seconds=20 finished=20.04
# task=3 pid=31864 seconds=20 finished=40.04
```

**Reply by shikokuchuo** [reply]
Well I have gone ahead and implemented the custom URL ports as that was a logical extension anyway - you simply specify a vector of URLs rather than a single URL. This should hopefully remove one of your obstacles! (v0.7.2.9010)

**Reply by wlandau** [reply]
Thanks!

Do you think the NNG maintainers would be open to a workaround? (Maybe disallow "send" to a given server until that server has the chance to send a message back?)

**Reply by shikokuchuo** [reply]
I think from their perspective, there is a reasonable expectation that we would be able to handle that ourselves.

They have been open to changes and indeed implemented one that made async in nanonext much more efficient. But I think the onus is on us to make the case that it needs to be at the NNG level.

**Reply by shikokuchuo** [reply]
We might get around this issue by reversing the listener/dialer relationship for the active queue as dialers do not bind to local ports.
However we end up with (i) the client needing to know the URLs of each server it connects to and (ii) each server will need to open a port for incoming connection from the client.

**Reply by wlandau** [reply]
Interesting. What if there is a single shared client port that all the servers can dial into to send their IP addresses, and then the client dials back on those IPs to send tasks? That would be 1 occupied port local to the client and 1 occupied port local to each server, right? 

I wonder how @mschubert's [`clustermq`](https://github.com/mschubert/clustermq) package handles all this. It does not have the auto-scaling and server-level control that `mirai` has, but it does have [workers (servers) that dial back into the client](https://github.com/mschubert/clustermq/blob/118b69089d72402856c9c11718192de10840070c/R/worker.r#L7-L11), and the load balancing is excellent without needing an active queue daemon. Could it be because of differences between ZeroMQ or NNG, or the maybe the difference in intended scope between `clustermq` and `mirai`?

**Reply by shikokuchuo** [reply]
Seems possible, but I'd have to think on it more. 2 potential added benefits:
(i) we could move the control signals (exit, status) to this new inbound port so they do not get blocked
(ii) we could add authentication to the handshake process based on a pre-shared secret (as implemented by the `messenger()` function in nanonext).

But I would have thought the open port on the server would be the thornier issue? Configuration on [1,000] servers vs. one client?

**Reply by shikokuchuo** [reply]
> I wonder how @mschubert's [`clustermq`](https://github.com/mschubert/clustermq) package handles all this. It does not have the auto-scaling and server-level control that `mirai` has, but it does have [workers (servers) that dial back into the client](https://github.com/mschubert/clustermq/blob/118b69089d72402856c9c11718192de10840070c/R/worker.r#L7-L11), and the load balancing is excellent without needing an active queue daemon. Could it be because of differences between ZeroMQ or NNG, or the maybe the difference in intended scope between `clustermq` and `mirai`?

There is (obviously!) no advantage of zeroMQ over NNG. From a cursory inspection - ClusterMQ also listens to a block of addresses. And it obtains them in a seemingly unsafe way simply by sampling from a range of 6000-9999, here:
https://github.com/mschubert/clustermq/blob/68c03c85caa888b0b3d573f5332cb9bb7dde5665/R/util.r#L35
But as I think it controls both client and server side, this is done opaquely and not visible to the user.

**Reply by wlandau** [reply]
Wow, I had no idea. Now I am surprised I haven't run into obvious port conflicts yet, even after many years my colleagues and I have used {clustermq}.

As for your idea about having the client dial into the server, what effect would that have on #33? Would that make it possible  to successfully monitor connections even with many mirais in the queue?

**Reply by shikokuchuo** [reply]
So maybe it's not as big of an issue as you are currently anticipating?

It has no effect on #33 let's not conflate the issues. Obviously the idea means there is a free inbound port at the queue, which I suggested we use for dual purposes. But if we need #33 we can address that separately.

This does open up another possibility which is to have a client port which servers can dial into to obtain the actual port it should dial into (a handshake process). This would be more-user friendly in only requiring one client URL, but it does require one extra port and masks the actual port usage...

**Reply by wlandau** [reply]
> So maybe it's not as big of an issue as you are currently anticipating?

It is starting to look that way. (And I just [posted a question](https://stackoverflow.com/questions/75638636/clarification-on-the-ephemeral-port-range-in-nng) asking about which ports NNG considers ephemeral.) I haven't decided whether I will add or keep schedules other than `mirai`, but in `crew` I now plan to build `mirai`-based worker launchers for both traditional clusters and cloud systems.

**Reply by wlandau** [reply]
As for #33: in `crew`'s autoscaling, I need some way to check which servers are currently running. Because of #33, I plan to go forward and do it on a case-by-case basis, e.g. `handle$is_alive()` for `callr::r_bg()` objects and `qstat` for SGE jobs. However, calling `qstat` too many times is a demand on the cluster's job scheduler, and the equivalent operation in the cloud is a call to the AWS API (which is cheap, but it's not free, and those API calls could could add up in cost). If there were an NNG/`mirai`-centric way to do this, then those costs could be avoided, and the check on the actual connection would be extra assurance that each server is healthy.

**Reply by shikokuchuo** [reply]
> > So maybe it's not as big of an issue as you are currently anticipating?
> 
> It is starting to look that way. (And I just [posted a question](https://stackoverflow.com/questions/75638636/clarification-on-the-ephemeral-port-range-in-nng) asking about which ports NNG considers ephemeral.) I haven't decided whether I will add or keep schedules other than `mirai`, but in `crew` I now plan to build `mirai`-based worker launchers for both traditional clusters and cloud systems.

The answer is that it is platform-dependent. NNG does not define this itself. Some operating systems you can even configure this port range system-wide. So a la wikipedia I guess.

**Reply by mschubert** [reply]
To add my 2 cents: the `clustermq` master listens on _one_ free port while the worker pool is active, so there is no "unsafe" behavior (maybe except for using a specific port that another program wants to use later on). So the `clustermq` solution is very similar to the solution in this discussion (use multiple incoming connections to one port, don't listen to a whole range; in fact it uses `REQ`/`ROUTER` sockets, similar to the `REQ`/`REP` that @shikokuchuo mentioned).

The main conceptual difference between `clustermq` and `mirai` as I see it, is that the former uses what @wlandau calls _persistent_ workers, while the latter uses _transient_ workers (that and the actual job submission).

**Reply by shikokuchuo** [reply]
> To add my 2 cents: the `clustermq` master listens on _one_ free port while the worker pool is active, so there is no "unsafe" behavior (maybe except for using a specific port that another program wants to use later on). So the `clustermq` solution is very similar to the solution in this discussion (use multiple incoming connections to one port, don't listen to a whole range; in fact it uses `REQ`/`ROUTER` sockets, similar to the `REQ`/`REP` that @shikokuchuo mentioned).
> 
> The main conceptual difference between `clustermq` and `mirai` as I see it, is that the former uses what @wlandau calls _persistent_ workers, while the latter uses _transient_ workers (that and the actual job submission).

@mschubert  thanks for making it so clear for us. Apologies, I only had a very quick glance through your code and came across where you sampled 'n' ports from a range and mistakenly came to the above conclusion. By 'unsafe', I only meant that there appears to be no guarantee that the selected port is free, in which case it would error - now thinking about it I am sure you handle this possibility in other parts of your code, which is probably why you sample a range of ports to fall back on.
